// @effect/opentelemetry/dist/NodeSdk.js
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
import * as Effect from "effect/Effect";
import { constant } from "effect/Function";
import * as Layer from "effect/Layer";
import { isNonEmpty } from "./internal/utilities.js";
import * as Logger from "./Logger.js";
import * as Metrics from "./Metrics.js";
import * as Resource from "./Resource.js";
import * as Tracer from "./Tracer.js";
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerTracerProvider = (processor, config) => Layer.effect(Tracer.OtelTracerProvider, Effect.gen(function* () {
  const resource = yield* Resource.Resource;
  return yield* Effect.acquireRelease(Effect.sync(() => {
    const provider = new NodeTracerProvider({
      ...(config ?? undefined),
      resource,
      spanProcessors: Array.isArray(processor) ? processor : [processor]
    });
    return provider;
  }), provider => Effect.promise(() => provider.forceFlush().then(() => provider.shutdown())).pipe(Effect.ignore, Effect.interruptible, Effect.timeoutOption(config?.shutdownTimeout ?? 3000)));
}));
/**
 * @since 1.0.0
 * @category Layers
 */
export const layer = evaluate => Layer.unwrap(Effect.gen(function* () {
  const config = yield* Effect.isEffect(evaluate) ? evaluate : Effect.sync(evaluate);
  const ResourceLive = Resource.layerFromEnv(config.resource && Resource.configToAttributes(config.resource));
  const TracerLive = isNonEmpty(config.spanProcessor) ? Layer.provide(Tracer.layer, layerTracerProvider(config.spanProcessor, {
    ...config.tracerConfig,
    shutdownTimeout: config.shutdownTimeout
  })) : Layer.empty;
  const MetricsLive = isNonEmpty(config.metricReader) ? Metrics.layer(constant(config.metricReader), {
    shutdownTimeout: config.shutdownTimeout,
    temporality: config.metricTemporality
  }) : Layer.empty;
  const LoggerLive = isNonEmpty(config.logRecordProcessor) ? Layer.provide(Logger.layer({
    mergeWithExisting: config.loggerMergeWithExisting
  }), Logger.layerLoggerProvider(config.logRecordProcessor, {
    ...config.loggerProviderConfig,
    shutdownTimeout: config.shutdownTimeout
  })) : Layer.empty;
  return Layer.mergeAll(TracerLive, MetricsLive, LoggerLive).pipe(Layer.provideMerge(ResourceLive));
}));
/**
 * @since 2.0.0
 * @category layer
 */
export const layerEmpty = Resource.layerEmpty;
