// @effect/opentelemetry/dist/Tracer.js
/**
 * @since 1.0.0
 */
import * as Otel from "@opentelemetry/api";
import * as OtelSemConv from "@opentelemetry/semantic-conventions";
import * as Cause from "effect/Cause";
import * as Context from "effect/Context";
import * as Effect from "effect/Effect";
import * as Exit from "effect/Exit";
import { constTrue, dual } from "effect/Function";
import * as Layer from "effect/Layer";
import * as Option from "effect/Option";
import * as Predicate from "effect/Predicate";
import * as Tracer from "effect/Tracer";
import { nanosToHrTime, recordToAttributes, unknownToAttributeValue } from "./internal/attributes.js";
import { Resource } from "./Resource.js";
// =============================================================================
// Service Definitions
// =============================================================================
/**
 * @since 1.0.0
 * @category Services
 */
export class OtelTracer extends /*#__PURE__*/Context.Service()("@effect/opentelemetry/Tracer") {}
/**
 * @since 1.0.0
 * @category Services
 */
export class OtelTracerProvider extends /*#__PURE__*/Context.Service()("@effect/opentelemetry/Tracer/OtelTracerProvider") {}
/**
 * @since 1.0.0
 * @category Services
 */
export class OtelTraceFlags extends /*#__PURE__*/Context.Service()("@effect/opentelemetry/Tracer/OtelTraceFlags") {}
/**
 * @since 1.0.0
 * @category Services
 */
export class OtelTraceState extends /*#__PURE__*/Context.Service()("@effect/opentelemetry/Tracer/OtelTraceState") {}
// =============================================================================
// Constructors
// =============================================================================
/**
 * @since 1.0.0
 * @category Constructors
 */
export const make = /*#__PURE__*/Effect.map(/*#__PURE__*/Effect.service(OtelTracer), tracer => Tracer.make({
  span(options) {
    return new OtelSpan(Otel.context, Otel.trace, tracer, options);
  },
  context(primitive, fiber) {
    const currentSpan = fiber.currentSpan;
    if (currentSpan === undefined) {
      return primitive["~effect/Effect/evaluate"](fiber);
    }
    return Otel.context.with(populateContext(Otel.context.active(), currentSpan), () => primitive["~effect/Effect/evaluate"](fiber));
  }
}));
/**
 * @since 1.0.0
 * @category Constructors
 */
export const makeExternalSpan = options => {
  const annotations = Context.mutate(Context.empty(), annotations => {
    let next = annotations;
    if (options.traceFlags !== undefined) {
      next = Context.add(next, OtelTraceFlags, options.traceFlags);
    }
    if (typeof options.traceState === "string") {
      try {
        next = Context.add(next, OtelTraceState, Otel.createTraceState(options.traceState));
      } catch {
        //
      }
    } else if (options.traceState) {
      next = Context.add(next, OtelTraceState, options.traceState);
    }
    return next;
  });
  return {
    _tag: "ExternalSpan",
    traceId: options.traceId,
    spanId: options.spanId,
    sampled: Predicate.isNotUndefined(options.traceFlags) ? isSampled(options.traceFlags) : true,
    annotations
  };
};
// =============================================================================
// Layers
// =============================================================================
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerGlobalProvider = /*#__PURE__*/Layer.sync(OtelTracerProvider, () => Otel.trace.getTracerProvider());
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerTracer = /*#__PURE__*/Layer.effect(OtelTracer, /*#__PURE__*/Effect.gen(function* () {
  const resource = yield* Resource;
  const provider = yield* OtelTracerProvider;
  return provider.getTracer(resource.attributes[OtelSemConv.ATTR_SERVICE_NAME], resource.attributes[OtelSemConv.ATTR_SERVICE_VERSION]);
}));
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerGlobalTracer = /*#__PURE__*/layerTracer.pipe(/*#__PURE__*/Layer.provide(layerGlobalProvider));
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerGlobal = /*#__PURE__*/Layer.effect(Tracer.Tracer, make).pipe(/*#__PURE__*/Layer.provideMerge(layerGlobalTracer));
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerWithoutOtelTracer = /*#__PURE__*/Layer.effect(Tracer.Tracer, make);
/**
 * @since 1.0.0
 * @category Layers
 */
export const layer = /*#__PURE__*/layerWithoutOtelTracer.pipe(/*#__PURE__*/Layer.provideMerge(layerTracer));
// =============================================================================
// Utilities / Combinators
// =============================================================================
const bigint1e6 = /*#__PURE__*/BigInt(1_000_000);
const bigint1e9 = /*#__PURE__*/BigInt(1_000_000_000);
/**
 * Get the current OpenTelemetry span.
 *
 * Works with both the official OpenTelemetry API (via `Tracer.layer`,
 * `NodeSdk.layer`, etc.) and the lightweight OTLP module (`OtlpTracer.layer`).
 *
 * When using OTLP, the returned span is a wrapper that conforms to the
 * OpenTelemetry `Span` interface.
 *
 * @since 1.0.0
 * @category accessors
 */
export const currentOtelSpan = /*#__PURE__*/Effect.clockWith(clock => Effect.map(Effect.currentSpan, span => OtelSpanTypeId in span ? span.span : makeOtelSpan(span, clock)));
const makeOtelSpan = (span, clock) => {
  const spanContext = {
    traceId: span.traceId,
    spanId: span.spanId,
    traceFlags: span.sampled ? Otel.TraceFlags.SAMPLED : Otel.TraceFlags.NONE,
    isRemote: false
  };
  let exit = Exit.void;
  const self = {
    spanContext: () => spanContext,
    setAttribute(key, value) {
      span.attribute(key, value);
      return self;
    },
    setAttributes(attributes) {
      for (const [key, value] of Object.entries(attributes)) {
        span.attribute(key, value);
      }
      return self;
    },
    addEvent(name) {
      let attributes = undefined;
      let startTime = undefined;
      if (arguments.length === 3) {
        attributes = arguments[1];
        startTime = arguments[2];
      } else {
        startTime = arguments[1];
      }
      span.event(name, convertOtelTimeInput(startTime, clock), attributes);
      return self;
    },
    addLink(link) {
      span.addLinks([{
        span: makeExternalSpan(link.context),
        attributes: link.attributes ?? {}
      }]);
      return self;
    },
    addLinks(links) {
      span.addLinks(links.map(link => ({
        span: makeExternalSpan(link.context),
        attributes: link.attributes ?? {}
      })));
      return self;
    },
    setStatus(status) {
      exit = Otel.SpanStatusCode.ERROR ? Exit.die(status.message ?? "Unknown error") : Exit.void;
      return self;
    },
    updateName: () => self,
    end(endTime) {
      const time = convertOtelTimeInput(endTime, clock);
      span.end(time, exit);
      return self;
    },
    isRecording: constTrue,
    recordException(exception, timeInput) {
      const time = convertOtelTimeInput(timeInput, clock);
      const cause = Cause.fail(exception);
      const error = Cause.prettyErrors(cause)[0];
      span.event(error.message, time, {
        "exception.type": error.name,
        "exception.message": error.message,
        "exception.stacktrace": error.stack ?? ""
      });
    }
  };
  return self;
};
const convertOtelTimeInput = (input, clock) => {
  if (input === undefined) {
    return clock.currentTimeNanosUnsafe();
  } else if (typeof input === "number") {
    return BigInt(Math.round(input * 1_000_000));
  } else if (input instanceof Date) {
    return BigInt(input.getTime()) * bigint1e6;
  }
  const [seconds, nanos] = input;
  return BigInt(seconds) * bigint1e9 + BigInt(nanos);
};
/**
 * Set the effect's parent span from the given opentelemetry `SpanContext`.
 *
 * This is handy when you set up OpenTelemetry outside of Effect and want to
 * attach to a parent span.
 *
 * @since 1.0.0
 * @category Propagation
 */
export const withSpanContext = /*#__PURE__*/dual(2, (self, spanContext) => Effect.withParentSpan(self, makeExternalSpan(spanContext)));
// =============================================================================
// Internals
// =============================================================================
const OtelSpanTypeId = "~@effect/opentelemetry/Tracer/OtelSpan";
const kindMap = {
  "internal": Otel.SpanKind.INTERNAL,
  "client": Otel.SpanKind.CLIENT,
  "server": Otel.SpanKind.SERVER,
  "producer": Otel.SpanKind.PRODUCER,
  "consumer": Otel.SpanKind.CONSUMER
};
/** @internal */
export class OtelSpan {
  [OtelSpanTypeId];
  _tag = "Span";
  name;
  kind;
  annotations;
  links;
  span;
  spanId;
  traceId;
  attributes = /*#__PURE__*/new Map();
  sampled;
  parent;
  status;
  constructor(contextApi, traceApi, tracer, options) {
    this[OtelSpanTypeId] = OtelSpanTypeId;
    this.name = options.name;
    this.annotations = options.annotations;
    this.links = options.links;
    this.kind = options.kind;
    const active = contextApi.active();
    this.parent = options.root !== true ? Option.orElse(options.parent, () => getOtelParent(traceApi, active, options.annotations)) : options.parent;
    this.span = tracer.startSpan(options.name, {
      startTime: nanosToHrTime(options.startTime),
      links: options.links.length > 0 ? options.links.map(link => ({
        context: makeSpanContext(link.span),
        attributes: recordToAttributes(link.attributes)
      })) : undefined,
      kind: kindMap[this.kind]
    }, Option.isSome(this.parent) ? populateContext(active, this.parent.value, options.annotations) : Otel.trace.deleteSpan(active));
    const spanContext = this.span.spanContext();
    this.spanId = spanContext.spanId;
    this.traceId = spanContext.traceId;
    this.status = {
      _tag: "Started",
      startTime: options.startTime
    };
    this.sampled = isSampled(spanContext.traceFlags);
  }
  attribute(key, value) {
    this.span.setAttribute(key, unknownToAttributeValue(value));
    this.attributes.set(key, value);
  }
  addLinks(links) {
    // oxlint-disable-next-line no-restricted-syntax
    this.links.push(...links);
    this.span.addLinks(links.map(link => ({
      context: makeSpanContext(link.span),
      attributes: recordToAttributes(link.attributes)
    })));
  }
  end(endTime, exit) {
    const hrTime = nanosToHrTime(endTime);
    this.status = {
      _tag: "Ended",
      endTime,
      exit,
      startTime: this.status.startTime
    };
    if (exit._tag === "Success") {
      this.span.setStatus({
        code: Otel.SpanStatusCode.OK
      });
    } else {
      if (Cause.hasInterruptsOnly(exit.cause)) {
        this.span.setStatus({
          code: Otel.SpanStatusCode.OK,
          message: Cause.pretty(exit.cause)
        });
        this.span.setAttribute("span.label", "⚠︎ Interrupted");
        this.span.setAttribute("status.interrupted", true);
      } else {
        const errors = Cause.prettyErrors(exit.cause);
        if (errors.length > 0) {
          for (const error of errors) {
            this.span.recordException(error, hrTime);
          }
          this.span.setStatus({
            code: Otel.SpanStatusCode.ERROR,
            message: errors[0].message
          });
        } else {
          // empty cause means no error
          this.span.setStatus({
            code: Otel.SpanStatusCode.OK
          });
        }
      }
    }
    this.span.end(hrTime);
  }
  event(name, startTime, attributes) {
    this.span.addEvent(name, attributes ? recordToAttributes(attributes) : undefined, nanosToHrTime(startTime));
  }
}
const isSampled = traceFlags => (traceFlags & Otel.TraceFlags.SAMPLED) === Otel.TraceFlags.SAMPLED;
const getOtelParent = (tracer, context, annotations) => {
  const otelParent = tracer.getSpan(context)?.spanContext();
  if (!otelParent) return Option.none();
  return Option.some(Tracer.externalSpan({
    spanId: otelParent.spanId,
    traceId: otelParent.traceId,
    sampled: (otelParent.traceFlags & 1) === 1,
    annotations
  }));
};
const makeSpanContext = (span, annotations) => {
  const traceFlags = makeTraceFlags(span, annotations);
  const traceState = makeTraceState(span, annotations);
  return {
    spanId: span.spanId,
    traceId: span.traceId,
    isRemote: span._tag === "ExternalSpan",
    traceFlags,
    traceState
  };
};
const makeTraceFlags = (span, annotations) => {
  let traceFlags;
  if (Predicate.isNotUndefined(annotations)) {
    traceFlags = extractTraceService(span, annotations, OtelTraceFlags);
    if (Predicate.isUndefined(traceFlags)) {
      traceFlags = Context.getOrUndefined(span.annotations, OtelTraceFlags);
    }
  }
  return traceFlags ?? Otel.TraceFlags.SAMPLED;
};
const makeTraceState = (span, annotations) => {
  let traceState;
  if (Predicate.isNotUndefined(annotations)) {
    traceState = extractTraceService(span, annotations, OtelTraceState);
    if (Predicate.isUndefined(traceState)) {
      traceState = Context.getOrUndefined(span.annotations, OtelTraceState);
    }
  }
  return traceState;
};
const extractTraceService = (parent, annotations, service) => {
  const instance = Context.getOrUndefined(annotations, service);
  if (Predicate.isNotUndefined(instance)) {
    return instance;
  }
  return Context.getOrUndefined(parent.annotations, service);
};
const populateContext = (context, span, annotations) => span instanceof OtelSpan ? Otel.trace.setSpan(context, span.span) : Otel.trace.setSpanContext(context, makeSpanContext(span, annotations));
