// @opentui/core/ffi-adapter.cjs
/**
 * node:ffi compatibility layer using koffi native module (CommonJS version)
 * 
 * Implements the node:ffi API surface used by @opentui/core's createNodeBackend(),
 * backed by the native koffi module (@koromix/koffi-openharmony-arm64).
 * 
 * Expected API from createNodeBackend():
 * 
 *   nodeFfi.dlopen(path, normalizedSymbols)
 *     -> { lib: { registerCallback(def, fn), unregisterCallback(ptr), close() },
 *          functions: { [name]: Function } }
 *   
 *   nodeFfi.getRawPointer(arrayBuffer) -> bigint
 *   nodeFfi.suffix -> ".so"
 *   nodeFfi.toArrayBuffer(ptr, offset, length) -> ArrayBuffer
 */

// ── Diagnostic logger: only prints when OPENTUI_DEBUG is set ──
const _diag = (() => {
  const enabled = !!process.env.OPENTUI_DEBUG;
  if (enabled) {
    return (...args) => console.log('[FFI-DIAG]', ...args);
  }
  return () => {};
})();

// Load koffi native module. We try the wrapped `koffi` package first (which wraps
// the native binary with JS helpers), then fall back to the arch-specific package.
function loadKoffi() {
  let resolved;
  try {
    resolved = require('koffi');
    _diag('resolved koffi via require("koffi")');
    return resolved;
  } catch {
    _diag('require("koffi") failed');
  }
  // HarmonyOS: try openharmony-arm64 first (process.platform === "openharmony")
  try {
    resolved = require('@koromix/koffi-openharmony-arm64');
    _diag('resolved koffi via require("@koromix/koffi-openharmony-arm64")');
    return resolved;
  } catch {
    _diag('require("@koromix/koffi-openharmony-arm64") failed');
  }
  throw new Error('Cannot find koffi native module');
}
const native = loadKoffi();

// ── Debug: inspect the koffi native binding ──
(function diagKoffi(nat) {
  const source = (() => {
    try { return require.resolve('koffi'); } catch {}
    try { return require.resolve('@koromix/koffi-openharmony-arm64'); } catch {}
    return '(unknown)';
  })();
  const keys = Object.keys(nat).sort();
  const type = typeof nat;
  const ctor = nat.constructor ? nat.constructor.name : 'N/A';
  _diag('koffi loaded from: %s', source);
  _diag('native type=%s ctor=%s keys=%j', type, ctor, keys);
  _diag('native.callback=%s', typeof nat.callback);
  _diag('native.proto=%s', typeof nat.proto);
  _diag('native.register=%s', typeof nat.register);
  _diag('native.unregister=%s', typeof nat.unregister);
  _diag('native.load=%s', typeof nat.load);

  // Try to find the actual .node binary path via Module._resolveFilename
  try {
    const Module = require('module');
    const resolvePaths = module.paths || [];
    _diag('module.paths count=%d', resolvePaths.length);
  } catch(e) {
    _diag('module.paths error: %s', e.message);
  }

  try {
    // For koffi wrapper packages, check the module's filename
    _diag('thisModule.filename=%s', module.filename);
  } catch(e) {
    _diag('filename error: %s', e.message);
  }

  // Dump require.resolve for all possible koffi native packages
  for (const pkgName of ['koffi', '@koromix/koffi-openharmony-arm64']) {
    try {
      _diag('require.resolve(%j)=%s', pkgName, require.resolve(pkgName));
    } catch {
      _diag('require.resolve(%j) FAILED', pkgName);
    }
  }
})(native);

// ============================================================
// Type Conversion
// ============================================================

/** Map from node:ffi type names to koffi type names */
const NODE_TO_KOFFI = {
  void:    "void",
  bool:    "bool",
  char:    "char",
  i8:      "int8",
  u8:      "uint8",
  int8:    "int8",
  uint8:   "uint8",
  i16:     "int16",
  u16:     "uint16",
  int16:   "int16",
  uint16:  "uint16",
  i32:     "int",
  u32:     "uint",
  int:     "int",
  int32:   "int",
  uint32:  "uint",
  i64:     "int64",
  u64:     "uint64",
  int64:   "int64",
  uint64:  "uint64",
  f32:     "float",
  float:   "float",
  f64:     "double",
  double:  "double",
  pointer: "void*",
  string:  "string",
  cstring: "string",
  buffer:  "void*",
  usize:   "size_t",
  size_t:  "size_t",
};

function toKoffiType(type) {
  if (!type) return "void";

  // Handle callback type definitions (objects like { result, parameters, ... })
  // For function binding purposes, treat callbacks as void* (function pointers)
  if (typeof type === "object" && type !== null) {
    return "void*";
  }

  // Handle callback type name strings like "void(int32,pointer)" that koffi
  // tries to parse as anonymous callback types. These fail because non-pointer
  // callback types can't be used as function parameters directly.
  const str = String(type);
  if (typeof type === "string" && /^[a-z_]\w*\(.*\)$/i.test(str)) {
    return "void*";
  }

  const lower = str.toLowerCase().trim();
  return NODE_TO_KOFFI[lower] || str;
}

// ============================================================
// Exports (matching createNodeBackend expectations)
// ============================================================

// ── dlopen 缓存：避免 @opentui/core 同步/异步双路径重复加载 ──
const dlopenCache = new Map();

/** @type {string} */
exports.suffix = ".so";

/**
 * Load a shared library and bind symbols.
 * 
 * @param {string} path - Path to .so file
 * @param {Object} symbols - { name: { result, parameters } }
 * @returns {{ lib: { registerCallback, unregisterCallback, close }, functions: Object }}
 */
exports.dlopen = function dlopen(path, symbols) {
  if (!path) throw new Error("Cannot dlopen null or undefined path");

  // 缓存命中则直接返回（防止 loadBackend / loadBackend2 重复初始化 koffi 类型表）
  const cacheKey = JSON.stringify({ path, symbolsKeys: Object.keys(symbols || {}) });
  if (dlopenCache.has(cacheKey)) {
    return dlopenCache.get(cacheKey);
  }

  const lib = native.load(path);
  const registeredCallbacks = [];

  /** @type {Object<string, Function>} */
  const functions = {};
  for (const [name, def] of Object.entries(symbols)) {
    const retTypeOrig = def.result;
    const paramTypesOrig = def.parameters || [];
    const retType = toKoffiType(retTypeOrig);
    const paramTypes = paramTypesOrig.map(toKoffiType);
    try {
      functions[name] = lib.func(name, retType, paramTypes);
    } catch (e) {
      throw new Error(`Failed to bind symbol '${name}': ${e.message}`);
    }
  }

  let cbCounter = 0;

  const libHandle = {
    /**
     * Register a JS function as a native callback.
     * Called by createNodeBackend() as: lib.registerCallback(normalizedDef, callback)
     * 
     * @param {{ result: string, parameters: string[] }} definition
     * @param {Function} callback
     * @returns {bigint} Native pointer to the callback
     */
    registerCallback(definition, callback) {
      if (definition.threadsafe) {
        throw new Error("Threadsafe callbacks not supported via koffi");
      }
      const ret = toKoffiType(definition.result);
      const params = (definition.parameters || []).map(toKoffiType);

      // Strategy: create a named callback type and register the JS function with it.
      // koffi v2/v3 API:
      //   native.callback(name, ret, params) — creates a NAMED callback type (compatible w/ register)
      //   native.callback(ret, params)       — anonymous callback type (2-arg form may not exist in all versions)
      //   native.register(fn, type)          — registers a JS callback, returns native pointer
      //   native.proto(name, ret, params)    — creates a function pointer type (NOT compatible w/ register!)
      //
      // We use the 3-argument callback(name, ret, params) form which is widely supported.

      const typeName = '_cb_' + process.pid + '_' + Date.now() + '_' + cbCounter++;
      const cbType = native.callback(typeName, ret, params);
      const ptr = native.register(callback, cbType);
      const entry = { ptr, closed: false };
      registeredCallbacks.push(entry);
      return BigInt(ptr);
    },

    /**
     * Unregister a native callback.
     * 
     * @param {bigint} ptr - Callback pointer
     */
    unregisterCallback(ptr) {
      const entry = registeredCallbacks.find(c => c.ptr === ptr);
      if (entry && !entry.closed) {
        entry.closed = true;
        native.unregister(ptr);
      }
    },

    /** Close the library and clean up all callbacks. */
    close() {
      for (const entry of registeredCallbacks) {
        if (!entry.closed) {
          entry.closed = true;
          native.unregister(entry.ptr);
        }
      }
      registeredCallbacks.length = 0;
    },
  };

  const result = { lib: libHandle, functions };
  dlopenCache.set(cacheKey, result);
  return result;
};

/**
 * Get the raw pointer address of an ArrayBuffer.
 * 
 * @param {ArrayBuffer} source
 * @returns {bigint}
 */
exports.getRawPointer = function getRawPointer(source) {
  return native.address(source);
};

/**
 * Read memory at a given pointer into an ArrayBuffer.
 * 
 * @param {bigint} pointer - Memory address
 * @param {number} offset - Byte offset from pointer
 * @param {number} length - Number of bytes to read
 * @returns {ArrayBuffer}
 */
// NOTE: Performance — each buf[i] = native.decode(...) is a N-API boundary
// call. For small reads (~512 bytes) this is ~50µs, acceptable for the current
// use case (log messages). For large buffers (100KB+) this would be ~10ms.
exports.toArrayBuffer = function toArrayBuffer(pointer, offset, length) {
  const addr = BigInt(pointer) + BigInt(offset ?? 0);
  if (addr < 0n) throw new Error("Invalid pointer value (negative)");

  const buf = Buffer.allocUnsafe(length);
  for (let i = 0; i < length; i++) {
    buf[i] = native.decode(addr + BigInt(i), "uint8");
  }

  const result = new ArrayBuffer(buf.length);
  new Uint8Array(result).set(buf);
  return result;
};

module.exports = exports;
