// @opentui/solid/jsx-runtime.js
// jsx-runtime.js for @opentui/solid
// Custom JSX runtime that uses @opentui/solid's custom reconciler
// instead of the DOM-based solid-js/h hyperscript.
//
// When esbuild compiles JSX with jsxImportSource: "@opentui/solid",
// it imports jsx/jsxs/Fragment from this file. The resolve-deps plugin
// in build-single-bundle.mjs maps @opentui/solid/jsx-runtime here.

import { createElement, spread, createComponent as _$createComponent } from "@opentui/solid"

function isComponent(type) {
  return typeof type === "function"
}

export function Fragment(props) {
  return props.children
}

/**
 * jsx / jsxs — create a TUI element node.
 *
 * For intrinsic elements (<box>, <text> etc.):
 *   - Uses @opentui/solid's custom createElement which creates a Renderable
 *     from the component catalogue (BoxRenderable, TextRenderable, etc.)
 *   - Properties are applied via spread() so reactive getters are tracked
 *
 * For custom components:
 *   - Returns a lazy function that Solid's insert() will evaluate via
 *     createComponent, preserving the reactive graph
 */
export function jsx(type, props, _key) {
  if (isComponent(type)) {
    // Custom Solid component — return as function so the universal
    // renderer's insert() can evaluate it reactively
    return () => _$createComponent(type, props || {})
  }

  // Intrinsic element — create a Renderable via the custom reconciler
  const el = createElement(type)

  if (props) {
    // Apply properties AND children via spread(). skipChildren=false
    // is essential — it lets spreadExpression call insertExpression(node,
    // props.children) which adds child Renderables to this element.
    spread(el, () => props, false)
  }

  return el
}

export function jsxs(type, props, _key) {
  return jsx(type, props)
}

export const jsxDEV = jsx
