// @effect/platform-node-shared/dist/NodeSocket.js
import * as Channel from "effect/Channel";
import * as Context from "effect/Context";
import * as Deferred from "effect/Deferred";
import * as Effect from "effect/Effect";
import * as FiberSet from "effect/FiberSet";
import * as Function from "effect/Function";
import { identity } from "effect/Function";
import * as Latch from "effect/Latch";
import * as Layer from "effect/Layer";
import * as Scope from "effect/Scope";
import * as Socket from "effect/unstable/socket/Socket";
import * as Net from "node:net";
/**
 * @since 1.0.0
 * @category re-exports
 */
export * as NodeWS from "ws";
/**
 * @since 1.0.0
 * @category tags
 */
export class NetSocket extends /*#__PURE__*/Context.Service()("@effect/platform-node/NodeSocket/NetSocket") {}
/**
 * @since 1.0.0
 * @category constructors
 */
export const makeNet = options => fromDuplex(Effect.contextWith(context => {
  let conn;
  return Effect.flatMap(Scope.addFinalizer(Context.get(context, Scope.Scope), Effect.sync(() => {
    if (!conn) return;
    if (conn.closed === false) {
      if ("destroySoon" in conn) {
        conn.destroySoon();
      } else {
        ;
        conn.destroy();
      }
    }
  })), () => Effect.callback(resume => {
    conn = Net.createConnection(options);
    conn.once("connect", () => {
      resume(Effect.succeed(conn));
    });
    conn.on("error", cause => {
      resume(Effect.fail(new Socket.SocketError({
        reason: new Socket.SocketOpenError({
          kind: "Unknown",
          cause
        })
      })));
    });
  }));
}), options);
/**
 * @since 1.0.0
 * @category constructors
 */
export const fromDuplex = (open, options) => Effect.withFiber(fiber => {
  let currentSocket;
  const latch = Latch.makeUnsafe(false);
  const openServices = fiber.context;
  const run = (handler, opts) => Effect.scopedWith(Effect.fnUntraced(function* (scope) {
    const fiberSet = yield* FiberSet.make().pipe(Scope.provide(scope));
    let conn = undefined;
    yield* Scope.addFinalizer(scope, Effect.sync(() => {
      if (!conn) return;
      conn.off("data", onData);
      conn.off("end", onEnd);
      conn.off("error", onError);
      conn.off("close", onClose);
    }));
    conn = yield* Scope.provide(open, scope).pipe(options?.openTimeout ? Effect.timeoutOrElse({
      duration: options.openTimeout,
      orElse: () => Effect.fail(new Socket.SocketError({
        reason: new Socket.SocketOpenError({
          kind: "Timeout",
          cause: new Error("Connection timed out")
        })
      }))
    }) : identity);
    conn.on("end", onEnd);
    conn.on("error", onError);
    conn.on("close", onClose);
    const run = yield* Effect.provideService(FiberSet.runtime(fiberSet)(), NetSocket, conn);
    conn.on("data", onData);
    currentSocket = conn;
    latch.openUnsafe();
    if (opts?.onOpen) {
      yield* opts.onOpen;
    }
    return yield* FiberSet.join(fiberSet);
    function onData(chunk) {
      const result = handler(chunk);
      if (Effect.isEffect(result)) {
        run(result);
      }
    }
    function onEnd() {
      Deferred.doneUnsafe(fiberSet.deferred, Effect.void);
    }
    function onError(cause) {
      Deferred.doneUnsafe(fiberSet.deferred, Effect.fail(new Socket.SocketError({
        reason: new Socket.SocketReadError({
          cause
        })
      })));
    }
    function onClose(hadError) {
      Deferred.doneUnsafe(fiberSet.deferred, Effect.fail(new Socket.SocketError({
        reason: new Socket.SocketCloseError({
          code: hadError ? 1006 : 1000
        })
      })));
    }
  })).pipe(Effect.updateContext(input => Context.merge(openServices, input)), Effect.onExit(() => Effect.sync(() => {
    latch.closeUnsafe();
    currentSocket = undefined;
  })));
  const write = chunk => latch.whenOpen(Effect.callback(resume => {
    const conn = currentSocket;
    if (Socket.isCloseEvent(chunk)) {
      conn.destroy(chunk.code > 1000 ? new Error(`closed with code ${chunk.code}`) : undefined);
      return resume(Effect.void);
    }
    currentSocket.write(chunk, cause => {
      resume(cause ? Effect.fail(new Socket.SocketError({
        reason: new Socket.SocketWriteError({
          cause: cause
        })
      })) : Effect.void);
    });
  }));
  const writer = Effect.acquireRelease(Effect.succeed(write), () => Effect.sync(() => {
    if (!currentSocket || currentSocket.writableEnded) return;
    currentSocket.end();
  }));
  return Effect.succeed(Socket.make({
    run,
    runRaw: run,
    writer
  }));
});
/**
 * @since 1.0.0
 * @category constructors
 */
export const makeNetChannel = options => Channel.unwrap(Effect.map(makeNet(options), Socket.toChannelWith()));
/**
 * @since 1.0.0
 * @category layers
 */
export const layerNet = /*#__PURE__*/Function.flow(makeNet, /*#__PURE__*/Layer.effect(Socket.Socket));
