// @effect/platform-node/dist/NodeHttpClient.js
/**
 * @since 1.0.0
 */
import * as Context from "effect/Context";
import * as Effect from "effect/Effect";
import { flow } from "effect/Function";
import * as Inspectable from "effect/Inspectable";
import * as Layer from "effect/Layer";
import * as Option from "effect/Option";
import { pipeArguments } from "effect/Pipeable";
import * as Stream from "effect/Stream";
import * as Cookies from "effect/unstable/http/Cookies";
import * as Headers from "effect/unstable/http/Headers";
import * as Client from "effect/unstable/http/HttpClient";
import * as Error from "effect/unstable/http/HttpClientError";
import * as Response from "effect/unstable/http/HttpClientResponse";
import * as IncomingMessage from "effect/unstable/http/HttpIncomingMessage";
import * as UrlParams from "effect/unstable/http/UrlParams";
import * as Http from "node:http";
import * as Https from "node:https";
import { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";
import { NodeHttpIncomingMessage } from "./NodeHttpIncomingMessage.js";
import * as NodeSink from "./NodeSink.js";
import * as NodeStream from "./NodeStream.js";
import * as Undici from "./Undici.js";
// -----------------------------------------------------------------------------
// Fetch
// -----------------------------------------------------------------------------
export {
/**
 * @since 1.0.0
 * @category Fetch
 */
Fetch,
/**
 * @since 1.0.0
 * @category Fetch
 */
layer as layerFetch,
/**
 * @since 1.0.0
 * @category Fetch
 */
RequestInit } from "effect/unstable/http/FetchHttpClient";
// -----------------------------------------------------------------------------
// Undici
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category Dispatcher
 */
export class Dispatcher extends /*#__PURE__*/Context.Service()("@effect/platform-node/NodeHttpClient/Dispatcher") {}
/**
 * @since 1.0.0
 * @category Dispatcher
 */
export const makeDispatcher = /*#__PURE__*/Effect.acquireRelease(/*#__PURE__*/Effect.sync(() => new Undici.Agent()), dispatcher => Effect.promise(() => dispatcher.destroy()));
/**
 * @since 1.0.0
 * @category Dispatcher
 */
export const layerDispatcher = /*#__PURE__*/Layer.effect(Dispatcher)(makeDispatcher);
/**
 * @since 1.0.0
 * @category Dispatcher
 */
export const dispatcherLayerGlobal = /*#__PURE__*/Layer.sync(Dispatcher)(() => Undici.getGlobalDispatcher());
/**
 * @since 1.0.0
 * @category undici
 */
export const UndiciOptions = /*#__PURE__*/Context.Reference("@effect/platform-node/NodeHttpClient/UndiciOptions", {
  defaultValue: () => ({})
});
/**
 * @since 1.0.0
 * @category undici
 */
export const makeUndici = /*#__PURE__*/Effect.gen(function* () {
  const dispatcher = yield* Dispatcher;
  return Client.make((request, url, signal, fiber) => convertBody(request.body).pipe(Effect.flatMap(body => Effect.tryPromise({
    try: () => dispatcher.request({
      ...fiber.getRef(UndiciOptions),
      signal,
      method: request.method,
      headers: request.headers,
      origin: url.origin,
      path: url.pathname + url.search + url.hash,
      body,
      // leave timeouts to Effect.timeout etc
      headersTimeout: 60 * 60 * 1000,
      bodyTimeout: 0
    }),
    catch: cause => new Error.HttpClientError({
      reason: new Error.TransportError({
        request,
        cause
      })
    })
  })), Effect.map(response => new UndiciResponse(request, response))));
});
function convertBody(body) {
  switch (body._tag) {
    case "Empty":
      {
        return Effect.succeed(null);
      }
    case "Uint8Array":
    case "Raw":
      {
        return Effect.succeed(body.body);
      }
    case "FormData":
      {
        return Effect.succeed(body.formData);
      }
    case "Stream":
      {
        return NodeStream.toReadable(body.stream);
      }
  }
}
function noopErrorHandler(_) {}
class UndiciResponse extends Inspectable.Class {
  [IncomingMessage.TypeId];
  [Response.TypeId];
  request;
  source;
  constructor(request, source) {
    super();
    this[IncomingMessage.TypeId] = IncomingMessage.TypeId;
    this[Response.TypeId] = Response.TypeId;
    this.request = request;
    this.source = source;
    source.body.on("error", noopErrorHandler);
  }
  get status() {
    return this.source.statusCode;
  }
  get statusText() {
    return undefined;
  }
  get headers() {
    return Headers.fromInput(this.source.headers);
  }
  cachedCookies;
  get cookies() {
    if (this.cachedCookies !== undefined) {
      return this.cachedCookies;
    }
    const header = this.source.headers["set-cookie"];
    return this.cachedCookies = header ? Cookies.fromSetCookie(header) : Cookies.empty;
  }
  get remoteAddress() {
    return Option.none();
  }
  get stream() {
    return NodeStream.fromReadable({
      evaluate: () => this.source.body,
      onError: cause => new Error.HttpClientError({
        reason: new Error.DecodeError({
          request: this.request,
          response: this,
          cause
        })
      })
    });
  }
  get json() {
    return Effect.flatMap(this.text, text => Effect.try({
      try: () => text === "" ? null : JSON.parse(text),
      catch: cause => new Error.HttpClientError({
        reason: new Error.DecodeError({
          request: this.request,
          response: this,
          cause
        })
      })
    }));
  }
  textBody;
  get text() {
    if (this.textBody) {
      return this.textBody;
    }
    this.textBody = Effect.tryPromise({
      try: () => this.source.body.text(),
      catch: cause => new Error.HttpClientError({
        reason: new Error.DecodeError({
          request: this.request,
          response: this,
          cause
        })
      })
    }).pipe(Effect.cached, Effect.runSync);
    this.arrayBufferBody = Effect.map(this.textBody, _ => new TextEncoder().encode(_).buffer);
    return this.textBody;
  }
  get urlParamsBody() {
    return Effect.flatMap(this.text, _ => Effect.try({
      try: () => UrlParams.fromInput(new URLSearchParams(_)),
      catch: cause => new Error.HttpClientError({
        reason: new Error.DecodeError({
          request: this.request,
          response: this,
          cause
        })
      })
    }));
  }
  formDataBody;
  get formData() {
    return this.formDataBody ??= Effect.tryPromise({
      try: () => this.source.body.formData(),
      catch: cause => new Error.HttpClientError({
        reason: new Error.DecodeError({
          request: this.request,
          response: this,
          cause
        })
      })
    }).pipe(Effect.cached, Effect.runSync);
  }
  arrayBufferBody;
  get arrayBuffer() {
    if (this.arrayBufferBody) {
      return this.arrayBufferBody;
    }
    this.arrayBufferBody = Effect.tryPromise({
      try: () => this.source.body.arrayBuffer(),
      catch: cause => new Error.HttpClientError({
        reason: new Error.DecodeError({
          request: this.request,
          response: this,
          cause
        })
      })
    }).pipe(Effect.cached, Effect.runSync);
    this.textBody = Effect.map(this.arrayBufferBody, _ => new TextDecoder().decode(_));
    return this.arrayBufferBody;
  }
  toJSON() {
    return IncomingMessage.inspect(this, {
      _id: "effect/http/HttpClientResponse",
      request: this.request.toJSON(),
      status: this.status
    });
  }
  pipe() {
    return pipeArguments(this, arguments);
  }
}
/**
 * @since 1.0.0
 * @category Undici
 */
export const layerUndiciNoDispatcher = /*#__PURE__*/Client.layerMergedContext(makeUndici);
/**
 * @since 1.0.0
 * @category Undici
 */
export const layerUndici = /*#__PURE__*/Layer.provide(layerUndiciNoDispatcher, layerDispatcher);
// -----------------------------------------------------------------------------
// node:http
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category HttpAgent
 */
export class HttpAgent extends /*#__PURE__*/Context.Service()("@effect/platform-node/NodeHttpClient/HttpAgent") {}
/**
 * @since 1.0.0
 * @category HttpAgent
 */
export const makeAgent = options => Effect.zipWith(Effect.acquireRelease(Effect.sync(() => new Http.Agent(options)), agent => Effect.sync(() => agent.destroy())), Effect.acquireRelease(Effect.sync(() => new Https.Agent(options)), agent => Effect.sync(() => agent.destroy())), (http, https) => ({
  http,
  https
}));
/**
 * @since 1.0.0
 * @category HttpAgent
 */
export const layerAgentOptions = /*#__PURE__*/flow(makeAgent, /*#__PURE__*/Layer.effect(HttpAgent));
/**
 * @since 1.0.0
 * @category HttpAgent
 */
export const layerAgent = /*#__PURE__*/layerAgentOptions();
/**
 * @since 1.0.0
 * @category node:http
 */
export const makeNodeHttp = /*#__PURE__*/Effect.gen(function* () {
  const agent = yield* HttpAgent;
  return Client.make((request, url, signal) => {
    const nodeRequest = url.protocol === "https:" ? Https.request(url, {
      agent: agent.https,
      method: request.method,
      headers: request.headers,
      signal
    }) : Http.request(url, {
      agent: agent.http,
      method: request.method,
      headers: request.headers,
      signal
    });
    return Effect.forkChild(sendBody(nodeRequest, request, request.body)).pipe(Effect.flatMap(() => waitForResponse(nodeRequest, request)), Effect.map(_ => new NodeHttpResponse(request, _)));
  });
});
const sendBody = (nodeRequest, request, body) => Effect.suspend(() => {
  switch (body._tag) {
    case "Empty":
      {
        nodeRequest.end();
        return waitForFinish(nodeRequest, request);
      }
    case "Uint8Array":
    case "Raw":
      {
        nodeRequest.end(body.body);
        return waitForFinish(nodeRequest, request);
      }
    case "FormData":
      {
        const response = new globalThis.Response(body.formData);
        response.headers.forEach((value, key) => {
          nodeRequest.setHeader(key, value);
        });
        return Effect.tryPromise({
          try: () => pipeline(Readable.fromWeb(response.body), nodeRequest),
          catch: cause => new Error.HttpClientError({
            reason: new Error.TransportError({
              request,
              cause
            })
          })
        });
      }
    case "Stream":
      {
        return Stream.run(Stream.mapError(body.stream, cause => new Error.HttpClientError({
          reason: new Error.EncodeError({
            request,
            cause
          })
        })), NodeSink.fromWritable({
          evaluate: () => nodeRequest,
          onError: cause => new Error.HttpClientError({
            reason: new Error.TransportError({
              request,
              cause
            })
          })
        }));
      }
  }
});
const waitForResponse = (nodeRequest, request) => Effect.callback(resume => {
  function onError(cause) {
    resume(Effect.fail(new Error.HttpClientError({
      reason: new Error.TransportError({
        request,
        cause
      })
    })));
  }
  nodeRequest.on("error", onError);
  function onResponse(response) {
    nodeRequest.off("error", onError);
    resume(Effect.succeed(response));
  }
  nodeRequest.on("upgrade", onResponse);
  nodeRequest.on("response", onResponse);
  return Effect.sync(() => {
    nodeRequest.off("error", onError);
    nodeRequest.off("upgrade", onResponse);
    nodeRequest.off("response", onResponse);
  });
});
const waitForFinish = (nodeRequest, request) => Effect.callback(resume => {
  function onError(cause) {
    resume(Effect.fail(new Error.HttpClientError({
      reason: new Error.TransportError({
        request,
        cause
      })
    })));
  }
  nodeRequest.once("error", onError);
  function onFinish() {
    nodeRequest.off("error", onError);
    resume(Effect.void);
  }
  nodeRequest.once("finish", onFinish);
  return Effect.sync(() => {
    nodeRequest.off("error", onError);
    nodeRequest.off("finish", onFinish);
  });
});
class NodeHttpResponse extends NodeHttpIncomingMessage {
  [Response.TypeId];
  request;
  constructor(request, source) {
    super(source, cause => new Error.HttpClientError({
      reason: new Error.DecodeError({
        request,
        response: this,
        cause
      })
    }));
    this[Response.TypeId] = Response.TypeId;
    this.request = request;
  }
  get status() {
    return this.source.statusCode;
  }
  cachedCookies;
  get cookies() {
    if (this.cachedCookies !== undefined) {
      return this.cachedCookies;
    }
    const header = this.source.headers["set-cookie"];
    return this.cachedCookies = header ? Cookies.fromSetCookie(header) : Cookies.empty;
  }
  get formData() {
    return Effect.tryPromise({
      try: () => {
        const init = {
          headers: new globalThis.Headers(this.source.headers)
        };
        if (this.source.statusCode) {
          init.status = this.source.statusCode;
        }
        if (this.source.statusMessage) {
          init.statusText = this.source.statusMessage;
        }
        return new globalThis.Response(Readable.toWeb(this.source), init).formData();
      },
      catch: this.onError
    });
  }
  toJSON() {
    return IncomingMessage.inspect(this, {
      _id: "effect/http/HttpClientResponse",
      request: this.request.toJSON(),
      status: this.status
    });
  }
  pipe() {
    return pipeArguments(this, arguments);
  }
}
/**
 * @since 1.0.0
 * @category node:http
 */
export const layerNodeHttpNoAgent = /*#__PURE__*/Client.layerMergedContext(makeNodeHttp);
/**
 * @since 1.0.0
 * @category node:http
 */
export const layerNodeHttp = /*#__PURE__*/Layer.provide(layerNodeHttpNoAgent, layerAgent);
