// @effect/platform-node/dist/NodeMultipart.js
/**
 * @since 1.0.0
 */
import * as Effect from "effect/Effect";
import * as Inspectable from "effect/Inspectable";
import * as Stream from "effect/Stream";
import * as Multipart from "effect/unstable/http/Multipart";
import * as MP from "effect/unstable/http/Multipasta/Node";
import * as NFS from "node:fs";
import * as NodeStreamP from "node:stream/promises";
import * as NodeStream from "./NodeStream.js";
/**
 * @since 1.0.0
 * @category constructors
 */
export const stream = (source, headers) => Multipart.makeConfig(headers).pipe(Effect.map(config => NodeStream.fromReadable({
  evaluate() {
    const parser = MP.make(config);
    source.pipe(parser);
    return parser;
  },
  onError: error => convertError(error)
})), Stream.unwrap, Stream.map(convertPart));
/**
 * @since 1.0.0
 * @category constructors
 */
export const persisted = (source, headers) => Multipart.toPersisted(stream(source, headers), (path, file) => Effect.tryPromise({
  try: signal => NodeStreamP.pipeline(file.file, NFS.createWriteStream(path), {
    signal
  }),
  catch: cause => Multipart.MultipartError.fromReason("InternalError", cause)
}));
/**
 * @since 1.0.0
 */
export const fileToReadable = file => file.file;
// -----------------------------------------------------------------------------
// Internal
// -----------------------------------------------------------------------------
const convertPart = part => part._tag === "Field" ? new FieldImpl(part.info, part.value) : new FileImpl(part);
class PartBase extends Inspectable.Class {
  [Multipart.TypeId];
  constructor() {
    super();
    this[Multipart.TypeId] = Multipart.TypeId;
  }
}
class FieldImpl extends PartBase {
  _tag = "Field";
  key;
  contentType;
  value;
  constructor(info, value) {
    super();
    this.key = info.name;
    this.contentType = info.contentType;
    this.value = MP.decodeField(info, value);
  }
  toJSON() {
    return {
      _id: "@effect/platform/Multipart/Part",
      _tag: "Field",
      key: this.key,
      value: this.value,
      contentType: this.contentType
    };
  }
}
class FileImpl extends PartBase {
  _tag = "File";
  key;
  name;
  contentType;
  content;
  contentEffect;
  file;
  constructor(file) {
    super();
    this.file = file;
    this.key = file.info.name;
    this.name = file.filename ?? file.info.name;
    this.contentType = file.info.contentType;
    this.content = NodeStream.fromReadable({
      evaluate: () => file,
      onError: cause => Multipart.MultipartError.fromReason("InternalError", cause)
    });
    this.contentEffect = NodeStream.toUint8Array(() => file, {
      onError: cause => Multipart.MultipartError.fromReason("InternalError", cause)
    });
  }
  toJSON() {
    return {
      _id: "@effect/platform/Multipart/Part",
      _tag: "File",
      key: this.key,
      name: this.name,
      contentType: this.contentType
    };
  }
}
function convertError(cause) {
  switch (cause._tag) {
    case "ReachedLimit":
      {
        switch (cause.limit) {
          case "MaxParts":
            {
              return Multipart.MultipartError.fromReason("TooManyParts", cause);
            }
          case "MaxFieldSize":
            {
              return Multipart.MultipartError.fromReason("FieldTooLarge", cause);
            }
          case "MaxPartSize":
            {
              return Multipart.MultipartError.fromReason("FileTooLarge", cause);
            }
          case "MaxTotalSize":
            {
              return Multipart.MultipartError.fromReason("BodyTooLarge", cause);
            }
        }
      }
    default:
      {
        return Multipart.MultipartError.fromReason("Parse", cause);
      }
  }
}
