// @effect/platform-node/dist/NodeWorker.js
/**
 * @since 1.0.0
 */
import * as Deferred from "effect/Deferred";
import * as Effect from "effect/Effect";
import * as Exit from "effect/Exit";
import * as Layer from "effect/Layer";
import * as Scope from "effect/Scope";
import * as Worker from "effect/unstable/workers/Worker";
import { WorkerError, WorkerReceiveError } from "effect/unstable/workers/WorkerError";
/**
 * @since 1.0.0
 * @category layers
 */
export const layerPlatform = /*#__PURE__*/Layer.succeed(Worker.WorkerPlatform)(/*#__PURE__*/Worker.makePlatform()({
  setup({
    scope,
    worker
  }) {
    const exitDeferred = Deferred.makeUnsafe();
    const thing = "postMessage" in worker ? {
      postMessage(msg, t) {
        worker.postMessage(msg, t);
      },
      kill: () => worker.terminate(),
      worker
    } : {
      postMessage(msg, _) {
        worker.send(msg);
      },
      kill: () => worker.kill("SIGKILL"),
      worker
    };
    worker.on("exit", () => {
      Deferred.doneUnsafe(exitDeferred, Exit.void);
    });
    return Effect.as(Scope.addFinalizer(scope, Effect.suspend(() => {
      thing.postMessage([1]);
      return Deferred.await(exitDeferred);
    }).pipe(Effect.timeout(5000), Effect.catchCause(() => Effect.sync(() => thing.kill())))), thing);
  },
  listen({
    deferred,
    emit,
    port
  }) {
    port.worker.on("message", message => {
      emit(message);
    });
    port.worker.on("messageerror", cause => {
      Deferred.doneUnsafe(deferred, new WorkerError({
        reason: new WorkerReceiveError({
          message: "An messageerror event was emitted",
          cause
        })
      }));
    });
    port.worker.on("error", cause => {
      Deferred.doneUnsafe(deferred, new WorkerError({
        reason: new WorkerReceiveError({
          message: "An error event was emitted",
          cause
        })
      }));
    });
    port.worker.on("exit", code => {
      Deferred.doneUnsafe(deferred, new WorkerError({
        reason: new WorkerReceiveError({
          message: "The worker has exited with code: " + code
        })
      }));
    });
    return Effect.void;
  }
}));
/**
 * @since 1.0.0
 * @category layers
 */
export const layer = spawn => Layer.merge(Worker.layerSpawner(spawn), layerPlatform);
