// @effect/platform-node/dist/NodeHttpServer.js
/**
 * @since 1.0.0
 */
import * as Cause from "effect/Cause";
import * as Config from "effect/Config";
import * as Context from "effect/Context";
import * as Duration from "effect/Duration";
import * as Effect from "effect/Effect";
import * as Fiber from "effect/Fiber";
import { flow } from "effect/Function";
import * as Latch from "effect/Latch";
import * as Layer from "effect/Layer";
import * as Scope from "effect/Scope";
import * as Stream from "effect/Stream";
import * as Cookies from "effect/unstable/http/Cookies";
import * as Etag from "effect/unstable/http/Etag";
import * as FetchHttpClient from "effect/unstable/http/FetchHttpClient";
import * as HttpEffect from "effect/unstable/http/HttpEffect";
import * as HttpIncomingMessage from "effect/unstable/http/HttpIncomingMessage";
import * as HttpServer from "effect/unstable/http/HttpServer";
import { causeResponse, ClientAbort, HttpServerError, RequestParseError, ResponseError, ServeError } from "effect/unstable/http/HttpServerError";
import * as Request from "effect/unstable/http/HttpServerRequest";
import { HttpServerRequest } from "effect/unstable/http/HttpServerRequest";
import * as Socket from "effect/unstable/socket/Socket";
import * as Http from "node:http";
import { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";
import { NodeHttpIncomingMessage } from "./NodeHttpIncomingMessage.js";
import * as NodeHttpPlatform from "./NodeHttpPlatform.js";
import * as NodeMultipart from "./NodeMultipart.js";
import * as NodeServices from "./NodeServices.js";
import { NodeWS } from "./NodeSocket.js";
/**
 * @since 1.0.0
 * @category constructors
 */
export const make = /*#__PURE__*/Effect.fnUntraced(function* (evaluate, options) {
  const scope = yield* Effect.scope;
  const server = evaluate();
  const shutdown = yield* Effect.callback(resume => {
    if (!server.listening) {
      return resume(Effect.void);
    }
    server.close(error => {
      if (error) {
        resume(Effect.die(error));
      } else {
        resume(Effect.void);
      }
    });
  }).pipe(Effect.cached);
  const preemptiveShutdown = options.disablePreemptiveShutdown ? Effect.void : Effect.timeoutOrElse(shutdown, {
    duration: options.gracefulShutdownTimeout ?? Duration.seconds(20),
    orElse: () => Effect.void
  });
  yield* Scope.addFinalizer(scope, shutdown);
  yield* Effect.callback(resume => {
    function onError(cause) {
      resume(Effect.fail(new ServeError({
        cause
      })));
    }
    server.on("error", onError);
    server.listen(options, () => {
      server.off("error", onError);
      resume(Effect.void);
    });
  });
  const address = server.address();
  const wss = yield* Effect.acquireRelease(Effect.sync(() => new NodeWS.WebSocketServer({
    noServer: true
  })), wss => Effect.callback(resume => {
    wss.close(() => resume(Effect.void));
  })).pipe(Scope.provide(scope), Effect.cached);
  return HttpServer.make({
    address: typeof address === "string" ? {
      _tag: "UnixAddress",
      path: address
    } : {
      _tag: "TcpAddress",
      hostname: address.address === "::" ? "0.0.0.0" : address.address,
      port: address.port
    },
    serve: Effect.fnUntraced(function* (httpApp, middleware) {
      const serveScope = yield* Effect.scope;
      const scope = Scope.forkUnsafe(serveScope, "parallel");
      const handler = yield* makeHandler(httpApp, {
        middleware: middleware,
        scope
      });
      const upgradeHandler = yield* makeUpgradeHandler(wss, httpApp, {
        middleware: middleware,
        scope
      });
      yield* Scope.addFinalizerExit(serveScope, () => {
        server.off("request", handler);
        server.off("upgrade", upgradeHandler);
        return preemptiveShutdown;
      });
      server.on("request", handler);
      server.on("upgrade", upgradeHandler);
    })
  });
});
/**
 * @since 1.0.0
 * @category Handlers
 */
export const makeHandler = (httpEffect, options) => {
  const handled = HttpEffect.toHandled(httpEffect, handleResponse, options.middleware);
  return Effect.withFiber(parent => {
    const services = parent.context;
    return Effect.succeed(function handler(nodeRequest, nodeResponse) {
      const map = new Map(services.mapUnsafe);
      map.set(HttpServerRequest.key, new ServerRequestImpl(nodeRequest, nodeResponse));
      const fiber = Fiber.runIn(Effect.runForkWith(Context.makeUnsafe(map))(handled), options.scope);
      nodeResponse.on("close", () => {
        if (!nodeResponse.writableEnded) {
          fiber.interruptUnsafe(parent.id, ClientAbort.annotation);
        }
      });
    });
  });
};
/**
 * @since 1.0.0
 * @category Handlers
 */
export const makeUpgradeHandler = (lazyWss, httpEffect, options) => {
  const handledApp = HttpEffect.toHandled(httpEffect, handleResponse, options.middleware);
  return Effect.withFiber(parent => {
    const services = parent.context;
    return Effect.succeed(function handler(nodeRequest, socket, head) {
      let nodeResponse_ = undefined;
      const nodeResponse = () => {
        if (nodeResponse_ === undefined) {
          nodeResponse_ = new Http.ServerResponse(nodeRequest);
          nodeResponse_.assignSocket(socket);
          nodeResponse_.on("finish", () => {
            socket.end();
          });
        }
        return nodeResponse_;
      };
      const upgradeEffect = Socket.fromWebSocket(Effect.flatMap(lazyWss, wss => Effect.acquireRelease(Effect.callback(resume => wss.handleUpgrade(nodeRequest, socket, head, ws => {
        resume(Effect.succeed(ws));
      })), ws => Effect.sync(() => ws.close()))));
      const map = new Map(services.mapUnsafe);
      map.set(HttpServerRequest.key, new ServerRequestImpl(nodeRequest, nodeResponse, upgradeEffect));
      const fiber = Fiber.runIn(Effect.runForkWith(Context.makeUnsafe(map))(handledApp), options.scope);
      socket.on("close", () => {
        if (!socket.writableEnded) {
          fiber.interruptUnsafe(parent.id, ClientAbort.annotation);
        }
      });
    });
  });
};
class ServerRequestImpl extends NodeHttpIncomingMessage {
  [Request.TypeId];
  response;
  upgradeEffect;
  url;
  headersOverride;
  constructor(source, response, upgradeEffect, url = source.url, headersOverride, remoteAddressOverride) {
    super(source, cause => new HttpServerError({
      reason: new RequestParseError({
        request: this,
        cause
      })
    }), remoteAddressOverride);
    this[Request.TypeId] = Request.TypeId;
    this.response = response;
    this.upgradeEffect = upgradeEffect;
    this.url = url;
    this.headersOverride = headersOverride;
  }
  cachedCookies;
  get cookies() {
    if (this.cachedCookies) {
      return this.cachedCookies;
    }
    return this.cachedCookies = Cookies.parseHeader(this.headers.cookie ?? "");
  }
  get resolvedResponse() {
    return typeof this.response === "function" ? this.response() : this.response;
  }
  modify(options) {
    return new ServerRequestImpl(this.source, this.response, this.upgradeEffect, options.url ?? this.url, options.headers ?? this.headersOverride, "remoteAddress" in options ? options.remoteAddress : this.remoteAddressOverride);
  }
  get originalUrl() {
    return this.source.url;
  }
  get method() {
    return this.source.method.toUpperCase();
  }
  get headers() {
    this.headersOverride ??= this.source.headers;
    return this.headersOverride;
  }
  multipartEffect;
  get multipart() {
    if (this.multipartEffect) {
      return this.multipartEffect;
    }
    this.multipartEffect = Effect.runSync(Effect.cached(NodeMultipart.persisted(this.source, this.source.headers)));
    return this.multipartEffect;
  }
  get multipartStream() {
    return NodeMultipart.stream(this.source, this.source.headers);
  }
  get upgrade() {
    return this.upgradeEffect ?? Effect.fail(new HttpServerError({
      reason: new RequestParseError({
        request: this,
        description: "not an upgradeable ServerRequest"
      })
    }));
  }
  toString() {
    return `ServerRequest(${this.method} ${this.url})`;
  }
  toJSON() {
    return HttpIncomingMessage.inspect(this, {
      _id: "HttpServerRequest",
      method: this.method,
      url: this.originalUrl
    });
  }
}
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerServer = /*#__PURE__*/flow(make, /*#__PURE__*/Layer.effect(HttpServer.HttpServer));
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerHttpServices = /*#__PURE__*/Layer.mergeAll(NodeHttpPlatform.layer, Etag.layerWeak, NodeServices.layer);
/**
 * @since 1.0.0
 * @category Layers
 */
export const layer = (evaluate, options) => Layer.mergeAll(layerServer(evaluate, options), layerHttpServices);
/**
 * @since 1.0.0
 * @category Layers
 */
export const layerConfig = (evaluate, options) => Layer.mergeAll(Layer.effect(HttpServer.HttpServer)(Effect.flatMap(Config.unwrap(options), options => make(evaluate, options))), layerHttpServices);
/**
 * @since 1.0.0
 * @category Testing
 */
export const layerTest = /*#__PURE__*/HttpServer.layerTestClient.pipe(/*#__PURE__*/Layer.provide(/*#__PURE__*/Layer.fresh(FetchHttpClient.layer).pipe(/*#__PURE__*/Layer.provide(/*#__PURE__*/Layer.succeed(FetchHttpClient.RequestInit)({
  keepalive: false
})))), /*#__PURE__*/Layer.provideMerge(/*#__PURE__*/layer(Http.createServer, {
  port: 0
})));
// -----------------------------------------------------------------------------
// Internal
// -----------------------------------------------------------------------------
const handleResponse = (request, response) => {
  const nodeResponse = request.resolvedResponse;
  if (nodeResponse.writableEnded) {
    return Effect.void;
  }
  let headers = response.headers;
  if (!Cookies.isEmpty(response.cookies)) {
    headers = {
      ...headers
    };
    const toSet = Cookies.toSetCookieHeaders(response.cookies);
    if (headers["set-cookie"] !== undefined) {
      toSet.push(headers["set-cookie"]);
    }
    headers["set-cookie"] = toSet;
  }
  if (request.method === "HEAD") {
    nodeResponse.writeHead(response.status, headers);
    return Effect.callback(resume => {
      nodeResponse.end(() => resume(Effect.void));
    });
  }
  const body = response.body;
  switch (body._tag) {
    case "Empty":
      {
        nodeResponse.writeHead(response.status, headers);
        nodeResponse.end();
        return Effect.void;
      }
    case "Raw":
      {
        nodeResponse.writeHead(response.status, headers);
        if (typeof body.body === "object" && body.body !== null && "pipe" in body.body && typeof body.body.pipe === "function") {
          return Effect.tryPromise({
            try: signal => pipeline(body.body, nodeResponse, {
              signal,
              end: true
            }),
            catch: cause => new HttpServerError({
              reason: new ResponseError({
                request,
                response,
                description: "Error writing raw response",
                cause
              })
            })
          }).pipe(Effect.interruptible, Effect.tapCause(handleCause(nodeResponse, response)));
        }
        return Effect.callback(resume => {
          nodeResponse.end(body.body, () => resume(Effect.void));
        });
      }
    case "Uint8Array":
      {
        nodeResponse.writeHead(response.status, headers);
        // If the body is less than 1MB, we skip the callback
        if (body.body.length < 1024 * 1024) {
          nodeResponse.end(body.body);
          return Effect.void;
        }
        return Effect.callback(resume => {
          nodeResponse.end(body.body, () => resume(Effect.void));
        });
      }
    case "FormData":
      {
        return Effect.suspend(() => {
          const r = new globalThis.Response(body.formData);
          nodeResponse.writeHead(response.status, {
            ...headers,
            ...Object.fromEntries(r.headers)
          });
          return Effect.callback((resume, signal) => {
            Readable.fromWeb(r.body, {
              signal
            }).pipe(nodeResponse).on("error", cause => {
              resume(Effect.fail(new HttpServerError({
                reason: new ResponseError({
                  request,
                  response,
                  description: "Error writing FormData response",
                  cause
                })
              })));
            }).once("finish", () => {
              resume(Effect.void);
            });
          }).pipe(Effect.interruptible, Effect.tapCause(handleCause(nodeResponse, response)));
        });
      }
    case "Stream":
      {
        nodeResponse.writeHead(response.status, headers);
        const drainLatch = Latch.makeUnsafe();
        nodeResponse.on("drain", () => drainLatch.openUnsafe());
        return body.stream.pipe(Stream.orDie, Stream.runForEachArray(array => {
          let needDrain = false;
          for (let i = 0; i < array.length; i++) {
            const written = nodeResponse.write(array[i]);
            if (!written && !needDrain) {
              needDrain = true;
              drainLatch.closeUnsafe();
            } else if (written && needDrain) {
              needDrain = false;
            }
          }
          if (!needDrain) return Effect.void;
          return drainLatch.await;
        }), Effect.interruptible, Effect.matchCauseEffect({
          onSuccess: () => Effect.sync(() => nodeResponse.end()),
          onFailure: handleCause(nodeResponse, response)
        }));
      }
  }
};
const handleCause = (nodeResponse, originalResponse) => originalCause => Effect.flatMap(causeResponse(originalCause), ([response, cause]) => {
  const headersSent = nodeResponse.headersSent;
  if (!headersSent) {
    nodeResponse.writeHead(response.status);
  }
  if (!nodeResponse.writableEnded) {
    nodeResponse.end();
  }
  return Effect.failCause(headersSent ? Cause.combine(originalCause, Cause.die(originalResponse)) : cause);
});
